Mathlib/Order/Completion.lean

```diff
@@ -0,0 +1,183 @@
+abbrev DedekindCut := Concept α α (· ≤ ·)
+
+namespace DedekindCut
+
+/-- The left set of a Dedekind cut. This is an alias for `Concept.extent`. -/
+abbrev left (A : DedekindCut α) : Set α := A.extent
+
+/-- The right set of a Dedekind cut. This is an alias for `Concept.intent`. -/
+abbrev right (A : DedekindCut α) : Set α := A.intent
+
+/-- See `DedekindCut.ext'` for a version using the right set. -/
+@[ext] theorem ext {A B : DedekindCut α} (h : A.left = B.left) : A = B := Concept.ext h
+
+/-- See `DedekindCut.ext` for a version using the left set. -/
+theorem ext' {A B : DedekindCut α} (h : A.right = B.right) : A = B := Concept.ext' h
+
+@[simp]
+theorem upperBounds_left (A : DedekindCut α) : upperBounds A.left = A.right :=
+  A.upperPolar_extent
+
+@[simp]
+theorem lowerBounds_right (A : DedekindCut α) : lowerBounds A.right = A.left :=
+  A.lowerPolar_intent
+
+theorem image_left_subset_lowerBounds {β : Type*} [Preorder β] {f : α → β} (hf : Monotone f)
+    (A : DedekindCut α) : f '' A.left ⊆ lowerBounds (f '' A.right) := by
+  rintro _ ⟨x, hx, rfl⟩ _ ⟨y, hy, rfl⟩
+  exact hf <| rel_extent_intent hx hy
+
+theorem image_right_subset_upperBounds {β : Type*} [Preorder β] {f : α → β} (hf : Monotone f)
+    (A : DedekindCut α) : f '' A.right ⊆ upperBounds (f '' A.left) := by
+  rintro _ ⟨x, hx, rfl⟩ _ ⟨y, hy, rfl⟩
+  exact hf <| rel_extent_intent hy hx
+
+/-- Convert an element into its Dedekind cut (`Iic a`, `Ici a`). This map is order-preserving,
+though it is injective only on partial orders. -/
+def of (a : α) : DedekindCut α :=
+  (Concept.ofObject _ a).copy
+    (Iic a) (by ext; simpa [mem_lowerPolar_iff] using forall_ge_iff_le.symm)
+    (Ici a) (by ext; simp)
+
+@[simp] theorem left_of (a : α) : (of a).left = Iic a := rfl
+@[simp] theorem right_of (a : α) : (of a).right = Ici a := rfl
+
+@[simp] theorem ofObject_eq_of (a : α) : ofObject (· ≤ ·) a = of a := (copy_eq ..).symm
+@[simp] theorem ofAttribute_eq_of (a : α) : ofAttribute (· ≤ ·) a = of a := by ext; simp
+
+@[simp]
+theorem of_le_of {a b : α} : of a ≤ of b ↔ a ≤ b := by
+  simpa using ofObject_le_ofAttribute_iff (r := (· ≤ ·)) (a := a)
+
+@[simp]
+theorem of_lt_of {a b : β} : of a < of b ↔ a < b := by
+  simp [lt_iff_le_not_ge]
+
+@[simp]
+theorem of_inj {a b : β} : of a = of b ↔ a = b := by
+  simp [le_antisymm_iff]
+
+/-- `DedekindCut.of` as an `OrderEmbedding`. -/
+@[simps! apply]
+def ofEmbedding : β ↪o DedekindCut β where
+  toFun := of
+  inj' _ _ := of_inj.1
+  map_rel_iff' := of_le_of
+
+@[simp] theorem ofEmbedding_coe : ⇑(@ofEmbedding β _) = of := rfl
+
+@[simp]
+theorem of_sSup (A : DedekindCut γ) : of (sSup A.left) = A := by
+  apply ext'
+  ext
+  rw [right_of, mem_Ici, sSup_le_iff, ← upperBounds_left, mem_upperBounds]
+
+@[simp]
+theorem of_sInf (A : DedekindCut γ) : of (sInf A.right) = A := by
+  ext
+  rw [left_of, mem_Iic, le_sInf_iff, ← lowerBounds_right, mem_lowerBounds]
+
+/-- Any order embedding `β ↪o γ` into a complete lattice factors through `DedekindCut β`.
```

Comment:
Can you write this first sentence with less lean and more words? Maybe something
```suggestion
/-- Any order embedding from a partial order `β` to a complete lattice factors through the complete lattice
of Dedekind cuts of `β`.
```